### ISSUE TITLE
Support creating windows from existing handles

### ISSUE THREAD

**Issue by ArildF** [Ultz/Silk.NET#189]
# Summary of feature
As far as I can tell Silk.NET insists on owning all window management itself. This makes it unsuitable for integration into other applications using toolkits such as Windows Forms, WPF or similar.

I am not requesting that you implement support for all possible toolkits out there. However, if there was a way to point Silk.NET to an existing HWND (and similar abstractions for other windowing systems) and use that for rendering it would go a long way.

**Comment by Perksey** [comment]
Hi there! Thanks you for choosing Silk.NET and suggesting this.

> As far as I can tell Silk.NET insists on owning all window management itself.

This isn't strictly true. In all versions you have been able to tack on your own windowing system, albeit it was less trivial in older versions, but newer versions implement the new [Context API](https://github.com/Ultz/Silk.NET/blob/master/documentation/proposals/Proposal%20-%20Contexts.md), which allows you to implement a relatively small `IGLContext` interface and pass it along to `GL.GetApi(IGLContext)`. And heck, even if that's still too much, you can just pass along a lambda for a GetProcAddress delegate.

> However, if there was a way to point Silk.NET to an existing HWND (and similar abstractions for other windowing systems) and use that for rendering it would go a long way.

This sadly isn't something we can't easily support given that GLFW (currently our only first-party windowing backend) doesn't support creating windows from existing handles at the moment, and we won't be able to provide a windowing implementation without binding to each of the native APIs. However, this may change in Silk.NET 2.0 when the SDL backend gets in (as SDL has the `SDL_CreateWindowFrom` function which supports HWNDs iirc).

In conclusion, we'll be able to support this in the main library hopefully by the end of the year via SDL. If you need something to use today, I recommend whipping up a minimal set of WGL bindings so you can either implement `IGLContext` proper (which won't do anything special until 2.0's validation features are in) or just pass `wglGetProcAddress` to `GL.GetApi()` like so:

```cs
var gl = GL.GetApi(x => wglGetProcAddress(x));
```

**Comment by FrostByteGER** [comment]
Indeed this is a very important feature!
It would allow me to integrate my engine DLL into a Winforms editor application. I exclusively use Vulkan, so using OpenGL specific code is not an option...

**Comment by Perksey** [comment]
@FrostByteGER For Vulkan can’t you just use the KHR_win32_surface extension and pass in your winforms hwnd?

**Comment by FrostByteGER** [comment]
Ac

> 
> 
> @FrostByteGER For Vulkan can’t you just use the KHR_win32_surface extension and pass in your winforms hwnd?

Actually yes, that should work!
As a sidenote this requires me to abstract the Silk.Net integration in my engine

**Comment by FrostByteGER** [comment]
This was pushed back to 3.0? :( @Perksey

**Comment by Perksey** [comment]
That was a mistake, have moved to 2.0 but with the help wanted and good first issue because we don't know where this fits into the preview cycle or if it'll make initial release yet, as we have a lot of issues to work through between now and the new year.

**Comment by FrostByteGER** [comment]
Ah I see! I'll take a look into your fresh 2.0 Preview 2, maybe I can help a bit :)

**Comment by Perksey** [comment]
> Ah I see! I'll take a look into your fresh 2.0 Preview 2, maybe I can help a bit :)
Much appreciated! Should be a little easier now that we have the SDL backend!

**Comment by Perksey** [comment]
Resolved in #340. In 2.0 Preview 3, there'll be a `SdlWindowing.CreateFrom` method which allows you to pass in a native window handle such as a HWND. SDL is not capable of picking up on an existing OpenGL context, so you'll need to whip up your own `IGLContext` in order to get windowing's OpenGL features to work.

TL;DR give a HWND and maybe a context to `SdlWindowing.CreateFrom` and it'll give you an `IView`.

If you're on a desktop platform, the returned `IView` will probably also implement `IWindow`

**Comment by grogorick** [comment]
In case anyone else ends here (or here #1160) searching for a way to only use the OpenGL function bindings of Silk.NET (i.e., without the windowing system or context creation), the following helped me:
https://sharovarskyi.com/blog/posts/csharp-win32-opengl-silknet/

**Comment by Perksey** [comment]
Good find!
